proposals/p2240.md

```diff
@@ -0,0 +1,1091 @@
+```c++
+template <ConvertibleToString... Ts>
+std::string StrCat(const Ts&... params) {
+  std::string result;
+  result.reserve((params.Length() + ... + 0));
+  for... (auto param: params) {
+    result.append(param.ToString());
+  }
+  return result;
+}
+```
+
+</td>
+<td>
+
+With SE-0393 and SE-404
+
+```swift
+func StrCat<each T: ConvertibleToString>(_ param: repeat each T) -> String {
+  var len: Int64 = 0;
+  for param in repeat each param {
+    len += param.Length()
+  }
+  var result: String = ""
+  result.reserveCapacity(len)
+  for param in repeat each param {
+    result.append(param.ToString())
+  }
+  return result
+}
+```
+
+</td>
+</tr>
+</table>
+
+## Rationale
+
+Carbon needs variadics to effectively support
+[interoperation with and migration from C++](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code),
+where variadic templates are fairly common. Variadics also make code
+[easier to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write),
+because some APIs (such as `printf`) can't be naturally expressed in terms of a
+fixed number of parameters.
+
+Furthermore, Carbon needs to support _generic_ variadics for the same reasons it
+needs to support generic non-variadic functions: for example,
+definition-checking makes APIs
+[easier to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write),
+and [easier to evolve](/docs/project/goals.md#software-and-language-evolution).
+Furthermore, the language as a whole is easier to understand and write code in
+if separate features like variadics and generics compose in natural ways, rather
+than being mutually exclusive.
+
+Variadics are also important for supporting
+[performance-critical software](/docs/project/goals.md#performance-critical-software),
+because variadic APIs can be more efficient than their non-variadic
+counterparts. For example, `StrCat` is fundamentally more efficient than
+something like a chain of `operator+` calls on `std::string`, because it does
+not need to materialize a series of partial results, and it can pre-allocate a
+buffer large enough for the final result.
+
+This proposal deviates somewhat from the principle that
+[all APIs are library APIs](/docs/project/principles/library_apis_only.md)
+because it treats tuples and packs as effectively built-in. As the document for
+this principle notes, tuples may be an exception to the principle because it is
+very hard to see how we could define them as library types without circularity.
+Packs have much the same problem, since a type representing a pack would itself
+have to be a variadic type. The need for both types to deeply support symbolic
+computation adds some major challenges here as well. However, if tuples could
+ever be library types, they would certainly need to be defined variadically, so
+in that sense this proposal may actually support that principle.
+
+## Alternatives considered
```

Comment:
Issue: restrictive matching of symbolic packs to parameter lists.

The more I think about this option, the more I dislike it:
* It creates an unnecessary difference between the template and symbolic cases at the call boundary, opposite the direction of #2153 .
* It is an additional rule to explain to developers, and the rule needs to be pretty subtle in order to allow things like forwarding and destructuring.
* I'm pretty worried that a consequence of this subtle rule is that we are going to reject things that developers will expect to work and it will be really hard to give an error explaining what they did wrong.
* It means there is additional state associated with a function declaration around how the arguments are declared that needs to be modeled somehow. I'm still working out how much of an issue this is, but for example the things I'm thinking about are how this interacts with #2875 and functions actually being implementations of the `Call` interface for a type.

Furthermore, I feel like the alternative is pretty good: there are cases that we have to reject, but in those cases there is a pretty clear error we can give ("no single return type could be deduced", or whatever).